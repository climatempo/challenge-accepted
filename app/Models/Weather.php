<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class Weather extends Model
{
    //
}
