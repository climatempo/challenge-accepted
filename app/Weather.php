<?php

namespace App;

use App\Support\Model\SourceData;

class Weather extends SourceData
{
    /**
     * Get weather by locale id
     * @param int $locale_id
     * @return array|\Illuminate\Support\Collection
     */
    public function byLocale(int $locale_id)
    {
        return $this->data->filter(function ($item) use ($locale_id) {
            return (int) $item["locale"]["id"] === $locale_id;
        })->first();
    }
}