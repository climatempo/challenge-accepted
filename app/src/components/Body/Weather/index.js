import React from 'react';

import {
	Container,
	Header,
	Date,
	Text,
	Body,
	Icon,
	Value,
	Group,
} from './styles';

import max from '../../../assets/img/icons/upload.png';
import min from '../../../assets/img/icons/download.png';
import drop from '../../../assets/img/icons/raindrop.png';
import umbrella from '../../../assets/img/icons/umbrella.png';

export default function Weather({ date, text, temperature, rain }) {
	return (
		<Container>
			<Header>
				<Date>{date}</Date>
				<Text>{text}</Text>
			</Header>
			<Body>
				<Group>
					<Icon source={max} />
					<Value maxTemperature>{temperature.max}ºC</Value>
				</Group>

				<Group>
					<Icon source={min} />
					<Value minTemperature>{temperature.min}ºC</Value>
				</Group>

				<Group mt>
					<Icon source={drop} />
					<Value>{rain.precipitation}mm</Value>
				</Group>

				<Group mt>
					<Icon source={umbrella} />
					<Value>{rain.probability}%</Value>
				</Group>
			</Body>
		</Container>
	);
}
