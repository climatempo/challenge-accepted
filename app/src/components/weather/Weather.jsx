import React, { Component } from 'react'
import axios from 'axios'
import moment from 'moment'

const baseUrl = 'http://localhost:5000/weather'
const initialState = {
    localeWeather: [],
    id: null
}

export default class Weather extends Component {

    constructor(props) {
        super(props)
        this.state = {
            ...initialState
        }
    }

    componentDidMount() {
        axios.post(`${baseUrl}?id=`).then(resp => {
            this.setState({ localeWeather: resp.data })
        })
    }


    componentDidUpdate = (prevProps, prevState) => {
        if(prevState.id !== this.props.localeId && this.state.id !== this.props.localeId){
            this.setState({ id: this.props.localeId })
            if (this.props.localeId && this.props.localeId !== null) {
                axios.post(`${baseUrl}?id=${this.props.localeId}`).then(resp => {
                    this.setState({ localeWeather: [resp.data] })
                })
            } else {
                axios.post(`${baseUrl}?id=`).then(resp => {
                    this.setState({ localeWeather: resp.data })
                })
            }
        }
    }

    renderWeather() {
        return this.state.localeWeather.map(city => {
            return (
                <div className="city" key={city.locale.id}>
                    <h2 className="align-center"><i className="fa fa-map-marker"></i> Previsão para {city.locale.name} - {city.locale.state}</h2>
                    <section className="tiles">

                        {city.weather.map(day => {
                            let data = moment(day.date, "YYYY-MM-DD")
                            return <article className="style1" key={Date.parse(day.date)}>
                                <div className="dayCard">
                                    <div className="dayDateText">
                                        <strong className="dayDate">{data.format("DD/MM/YYYY")}</strong>
                                        <p className="dayText">{day.text}</p>
                                    </div>
                                    <div className="infos">
                                        <span className="tempMax"><i className="icon fa fa-thermometer-full" /> 
                                            {day.temperature.max}ºC</span>
                                        <span className="tempMin"><i className="icon fa fa-thermometer-1" /> 
                                            {day.temperature.min}ºC</span>
                                        <span className="rainPrep"><i className="icon fa fa-tint" /> 
                                            {day.rain.precipitation}mm</span>
                                        <span className="rainProb"><i className="icon fa fa-umbrella" /> 
                                            {day.rain.probability}%</span>
                                    </div>
                                </div>
                            </article>
                        })
                        }
                    </section>
                </div>
            )
        })
    }

    render() {
        return (
            <React.Fragment>
                {this.renderWeather()}
            </React.Fragment>

        )
    }
}