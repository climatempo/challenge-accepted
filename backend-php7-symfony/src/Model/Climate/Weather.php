<?php

namespace App\Model\Climate;

use App\Service\RedisCache;

class Weather
{
    private static $weathers = [];

    private static function getWeatherData()
    {
        if (empty(self::$weathers)) {
            try {
                self::$weathers = json_decode(file_get_contents(__DIR__ . '/weather.json'));
            } catch(Exception $err) {
                // TODO: add logger
            }
        }

        return self::$weathers;
    }

    public static function getWeatherByLocaleId(int $localeId)
    {
        $result = array_filter(self::getWeatherData(), function($weather) use ($localeId) {
            return $weather->locale->id === $localeId;
        });

        return array_values($result);
    }

}
