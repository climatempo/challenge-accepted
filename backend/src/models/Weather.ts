import { Column, Entity, PrimaryGeneratedColumn } from "typeorm";

@Entity({ name: "weathers" })
export class Weather {
  @PrimaryGeneratedColumn("uuid")
  id!: string;

  @Column({ name: "locale_id" })
  localeId!: string;

  @Column()
  date!: Date;

  @Column()
  text!: string;

  @Column({ name: "temperature_min" })
  temperatureMin!: number;

  @Column({ name: "temperature_max" })
  temperatureMax!: number;

  @Column({ name: "rain_probability" })
  rainProbability!: number;

  @Column({ name: "rain_precipitation" })
  rainPrecipitation!: number;
}
