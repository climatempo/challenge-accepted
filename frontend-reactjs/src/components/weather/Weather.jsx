import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Card from '../Card';

const Container = styled.div`
  padding: 0 20px;
  h3 {
    font-size: 1.5em;
    color: #262626;
  }
`;

const Weather = ({ weather }) => (
  <Container>
    {weather && (
      <>
        <h3>
          Previsão para&nbsp;
          {weather.locale && weather.locale.name}
          &nbsp;-&nbsp;
          {weather.locale && weather.locale.state}
        </h3>
        {weather.weather && weather.weather.map((w) => (
          <WeatherCard key={w.date} weather={w} />
        ))}
      </>
    )}
  </Container>
);

export default Weather;

Weather.propTypes = {
  weather: PropTypes.object,
};

const WeatherHeader = styled.div`
  background-color: #fff;
  height: 80px;
  padding: 10px;
  span {
    font-size: 1.2em;
  }
  small {
    margin: 5px 0;
  }
`;

const WeatherContent = styled.div`
  display: grid;
  grid-column-gap: 10px;
  grid-row-gap: 25px;
  grid-template-columns: auto auto;
  margin: auto;
  width: 100%;
`;

const WeatherCard = ({ weather }) => (
  <Card>
    <>
      <WeatherHeader>
        <span>{weather.formatedDate}</span>
        <small>{weather.text}</small>
      </WeatherHeader>
      <WeatherContent>
        <InfoWithIcon
          color="blue"
          label={`${weather.temperature.max}°C`}
          iconUrl={`${window.location.origin}/images/icons/upload.png`}
        />
        <InfoWithIcon
          color="red"
          label={`${weather.temperature.min}°C`}
          iconUrl={`${window.location.origin}/images/icons/download.png`}
        />
        <InfoWithIcon
          label={`${weather.rain.precipitation}mm`}
          iconUrl={`${window.location.origin}/images/icons/raindrop-close-up.png`}
        />
        <InfoWithIcon
          label={`${weather.rain.probability}%`}
          iconUrl={`${window.location.origin}/images/icons/protection-symbol-of-opened-umbrella-silhouette-under-raindrops.png`}
        />
      </WeatherContent>
    </>
  </Card>
);

WeatherCard.propTypes = {
  weather: PropTypes.object,
};

const InfoContainer = styled.div`
  flex-direction: row;
  justify-content: center;
  align-items: center;
  font-size: 1.8em;
`;

const Label = styled.div`
  padding: 0 10px;
  color: ${({ color }) => color};
`;

const InfoWithIcon = ({ color, label, iconUrl }) => (
  <InfoContainer>
    <div>
      <img src={iconUrl} alt="" style={{ width: '30px' }} />
    </div>
    <Label color={color}>{label}</Label>
  </InfoContainer>
);

InfoWithIcon.propTypes = {
  color: PropTypes.string,
  label: PropTypes.string,
  iconUrl: PropTypes.string,
};

const MessageContainer = styled.div`
  flex-grow: 1;
  justify-content: center;
  align-items: center;
  color: #b3b3b3;
  font-size: 1.7em;
`;

export const Message = () => (
  <MessageContainer>
    <img
      src={`${window.location.origin}/images/icons/search-solid.svg`}
      alt=""
      style={{ width: '100px' }}
    />
  </MessageContainer>
);
