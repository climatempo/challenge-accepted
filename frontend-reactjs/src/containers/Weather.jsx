import React, { Component } from 'react';
import moment from 'moment';

import RegionSearchApi from '../services/region-search';
import WeatherApi from '../services/weather';

import Autocomplete from '../components/Autocomplete';
import WeatherComponent, { Message } from '../components/weather/Weather';

import throttle from '../utils/throttle';

export default class Weather extends Component {
  constructor() {
    super();

    this.state = {
      text: '',
      regions: [],
      weather: null,
    };

    this.throttle = throttle(1000);
  }

  getWeather() {
    return new Promise((resolve) => {
      const { text, regions } = this.state;

      if (!regions.length) return;

      let region = regions.find((r) => r.name === text);
      region = region || regions.find((c) => c.name.match(new RegExp(text, 'gi'))); // TODO: Fix firefox bug with datalist

      WeatherApi.getByLocaleId(region.id).then((resp) => {
        const { weather } = resp.data;
        (weather.weather || []).forEach((w) => {
          w.formatedDate = moment(w.date, 'YYYY-MM-DD').format('DD/MM/YYYY');
        });

        this.setState({ weather });

        resolve(resp);
      }).catch((err) => {
        window.console.error(err);
      });
    });
  }

  search(value) {
    const { regions } = this.state;

    return new Promise((resolve) => {
      this.setState({ text: value }, () => {
        if (regions.find((r) => r.name === value)) return;
        if (value && value.length) {
          this.throttle(() => {
            RegionSearchApi.searchByName(value).then((resp) => {
              this.setState({ regions: resp.data.regions });
              resolve(resp);
            }).catch((err) => {
              window.console.error(err);
            });
          });
        }
      });
    });
  }

  render() {
    const { regions, text, weather } = this.state;
    return (
      <>
        <Autocomplete
          list={regions}
          value={text}
          onChange={({ target: { value } }) => this.search(value)}
          onSearch={() => this.getWeather()}
        />
        {weather ? (
          <WeatherComponent weather={weather} />
        ) : (
          <Message />
        )}
      </>
    );
  }
}
