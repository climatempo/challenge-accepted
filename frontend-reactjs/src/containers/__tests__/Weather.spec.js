import React from 'react';
import toJson from 'enzyme-to-json';

import { shallow } from '../../tests-setup';

jest.mock('../../services/region-search', () => ({
  searchByName: jest.fn().mockImplementation(() => Promise.resolve({
    data: { regions: [{ id: 1, name: 'foo' }] },
  })),
}));

const weatherResp = {
  period: {
    begin: '2017-02-01',
    end: '2017-02-07',
  },
  locale: {
    id: 3735,
    name: 'Osasco',
    state: 'SP',
    latitude: -23.532,
    longitude: -46.792
  },
  weather: [
    {
      date: '2017-02-01',
      text: 'Sol com muitas nuvens durante o dia. Períodos de nublado, com chuva a qualquer hora.',
      temperature: {
        min: 20,
        max: 28
      },
      rain: {
        probability: 60,
        precipitation: 20
      }
    }
  ]
};

jest.mock('../../services/weather', () => ({
  getByLocaleId: jest.fn().mockImplementation(() => Promise.resolve({
    data: {
      // TODO: babel-plugin-jest-hoist: The module factory of `jest.mock()` is not allowed to reference any out-of-scope variables.
      weather: {
        period: {
          begin: '2017-02-01',
          end: '2017-02-07',
        },
        locale: {
          id: 3735,
          name: 'Osasco',
          state: 'SP',
          latitude: -23.532,
          longitude: -46.792
        },
        weather: [
          {
            date: '2017-02-01',
            text: 'Sol com muitas nuvens durante o dia. Períodos de nublado, com chuva a qualquer hora.',
            temperature: {
              min: 20,
              max: 28
            },
            rain: {
              probability: 60,
              precipitation: 20
            }
          }
        ]
      },
    },
  })),
}));

import Weather from '../Weather';

jest.mock('axios');

const regions = [{ id: 1, name: 'foo' }];

describe('Weather Container', () => {
  let wrapper = null;

  beforeEach(() => {
    wrapper = shallow(<Weather />);
  });

  it('should correctly render', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should search region names', async () => {
    await wrapper.instance().search('foo');
    expect(wrapper.state().text).toEqual('foo');
    expect(wrapper.state().regions).toEqual(expect.arrayContaining(regions));
  });

  it('Should get weather', async () => {
    wrapper.setState({ text: 'foo', regions });
    await wrapper.instance().getWeather();
    expect(wrapper.state().text).toEqual('foo');
    expect(wrapper.state().weather).toMatchObject(weatherResp);
  });

});
