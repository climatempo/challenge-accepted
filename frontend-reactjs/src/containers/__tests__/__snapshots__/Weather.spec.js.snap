// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather Container should correctly render 1`] = `
<Fragment>
  <Autocomplete
    list={Array []}
    onChange={[Function]}
    onSearch={[Function]}
    value=""
  />
  <Message />
</Fragment>
`;
